ssl/ssl_sess.c

```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
     ctx->verify_stateless_cookie_cb = cb;
 }
 
+SSL_SESSION *SSL_get1_previous_client_session(SSL *s)
+{
+    SSL_SESSION *ret = NULL;
+    SSL_SESSION data;
+    uint32_t mode;
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(s);
+
+    if (sc == NULL)
+        return 0;
+
+    mode = sc->session_ctx->session_cache_mode;
+
+    /* Not client mode w/o server also */
+    if ((mode & SSL_SESS_CACHE_BOTH) != SSL_SESS_CACHE_CLIENT) {
+        ERR_raise(ERR_LIB_SSL, SSL_R_BAD_CACHE_MODE);
+        return NULL;
+    }
+
+    if ((mode & SSL_SESS_CACHE_NO_INTERNAL_LOOKUP) == 0) {
+        /* Set up the search parameters */
+        memset(&data, 0, sizeof(data));
+        data.cache_id = sc->cache_id;
+        data.cache_id_len = sc->cache_id_len;
+        data.sid_ctx_length = sc->sid_ctx_length;
+        if (sc->sid_ctx_length > 0)
+            memcpy(data.sid_ctx, sc->sid_ctx, sc->sid_ctx_length);
+
+        /* search */
+        if (!CRYPTO_THREAD_read_lock(sc->session_ctx->lock)) {
+            return NULL;
+        }
+        ret = lh_SSL_SESSION_retrieve(sc->session_ctx->sessions, &data);
+        if (ret != NULL)
+            SSL_SESSION_up_ref(ret);
+        CRYPTO_THREAD_unlock(sc->session_ctx->lock);
+
+        if (ret == NULL) {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_miss);
+        } else if (sess_timedout(ossl_time_now(), ret)) {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_timeout);
+            SSL_CTX_remove_session(sc->session_ctx, ret);
+            SSL_SESSION_free(ret);
+            ret = NULL;
+        } else {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_hit);
+        }
+    }
+
+    if (ret == NULL && sc->session_ctx->get_session_cb != NULL) {
+        int copy = 1;
+
+        ret = sc->session_ctx->get_session_cb(s, sc->cache_id,
+                                             sc->cache_id_len,
+                                             &copy);
```

Comment:
Should these also be checked for timeout, or is that solely up to the external cache?